In [14]:
import pandas as pd
import numpy as np

In [62]:
df = pd.read_csv('OnlineNewsPopularity.csv')

In [63]:
df.head()

url   timedelta  \
0  http://mashable.com/2013/01/07/amazon-instant-...       731.0   
1  http://mashable.com/2013/01/07/ap-samsung-spon...       731.0   
2  http://mashable.com/2013/01/07/apple-40-billio...       731.0   
3  http://mashable.com/2013/01/07/astronaut-notre...       731.0   
4   http://mashable.com/2013/01/07/att-u-verse-apps/       731.0   

    n_tokens_title   n_tokens_content   n_unique_tokens   n_non_stop_words  \
0             12.0              219.0          0.663594                1.0   
1              9.0              255.0          0.604743                1.0   
2              9.0              211.0          0.575130                1.0   
3              9.0              531.0          0.503788                1.0   
4             13.0             1072.0          0.415646                1.0   

    n_non_stop_unique_tokens   num_hrefs   num_self_hrefs   num_imgs  ...  \
0                   0.815385         4.0              2.0        1.0  ...   
1                   0.791946         3.0              1.0        1.0  ...   
2                   0.663866         3.0              1.0        1.0  ...   
3                   0.665635         9.0              0.0        1.0  ...   
4                   0.540890        19.0             19.0       20.0  ...   

    min_positive_polarity   max_positive_polarity   avg_negative_polarity  \
0                0.100000                     0.7               -0.350000   
1                0.033333                     0.7               -0.118750   
2                0.100000                     1.0               -0.466667   
3                0.136364                     0.8               -0.369697   
4                0.033333                     1.0               -0.220192   

    min_negative_polarity   max_negative_polarity   title_subjectivity  \
0                  -0.600               -0.200000             0.500000   
1                  -0.125               -0.100000             0.000000   
2                  -0.800               -0.133333             0.000000   
3                  -0.600               -0.166667             0.000000   
4                  -0.500               -0.050000             0.454545   

    title_sentiment_polarity   abs_title_subjectivity  \
0                  -0.187500                 0.000000   
1                   0.000000                 0.500000   
2                   0.000000                 0.500000   
3                   0.000000                 0.500000   
4                   0.136364                 0.045455   

    abs_title_sentiment_polarity   shares  
0                       0.187500      593  
1                       0.000000      711  
2                       0.000000     1500  
3                       0.000000     1200  
4                       0.136364      505  

[5 rows x 61 columns]

In [64]:
columns=[]
for i in df.columns:
    i=i.lstrip()
    columns.append(i)

In [65]:
df.columns=columns
df.isnull().sum().mean()

0.0

In [66]:
df=df[df['n_tokens_content']!=0]

In [67]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38463 entries, 0 to 39643
Data columns (total 61 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   url                            38463 non-null  object 
 1   timedelta                      38463 non-null  float64
 2   n_tokens_title                 38463 non-null  float64
 3   n_tokens_content               38463 non-null  float64
 4   n_unique_tokens                38463 non-null  float64
 5   n_non_stop_words               38463 non-null  float64
 6   n_non_stop_unique_tokens       38463 non-null  float64
 7   num_hrefs                      38463 non-null  float64
 8   num_self_hrefs                 38463 non-null  float64
 9   num_imgs                       38463 non-null  float64
 10  num_videos                     38463 non-null  float64
 11  average_token_length           38463 non-null  float64
 12  num_keywords                   38463 non-null 

In [68]:
df["category"] = "lifestyle"
df.loc[df["data_channel_is_tech"] == 1, "category"] = "tech"
df.loc[df["data_channel_is_socmed"] == 1, "category"] = "socmed"
df.loc[df["data_channel_is_bus"] == 1, "category"] = "business"
df.loc[df["data_channel_is_entertainment"] == 1, "category"] = "Entertainment"
df.loc[df["data_channel_is_world"] == 1, "category"] = "World"

In [69]:
df['Day']='Sunday'
# Filter the data for Monday
df.loc[df['weekday_is_monday'] == 1, 'Day']='Monday'
df.loc[df['weekday_is_tuesday'] == 1, 'Day']='Tuesday'
df.loc[df['weekday_is_wednesday'] == 1, 'Day']='Wednesday'
df.loc[df['weekday_is_thursday'] == 1, 'Day']='Thursday'
df.loc[df['weekday_is_friday'] == 1, 'Day']='Friday'
df.loc[df['weekday_is_saturday'] == 1, 'Day']='Saturday'

In [70]:
df=df.drop(["weekday_is_monday","weekday_is_tuesday","weekday_is_wednesday","weekday_is_thursday",
                  "weekday_is_friday","weekday_is_saturday","weekday_is_sunday","is_weekend",                  
                  "data_channel_is_lifestyle","data_channel_is_entertainment","data_channel_is_bus",
                  "data_channel_is_socmed","data_channel_is_tech","data_channel_is_world"],axis=1)

In [71]:
df.head()

url  timedelta  \
0  http://mashable.com/2013/01/07/amazon-instant-...      731.0   
1  http://mashable.com/2013/01/07/ap-samsung-spon...      731.0   
2  http://mashable.com/2013/01/07/apple-40-billio...      731.0   
3  http://mashable.com/2013/01/07/astronaut-notre...      731.0   
4   http://mashable.com/2013/01/07/att-u-verse-apps/      731.0   

   n_tokens_title  n_tokens_content  n_unique_tokens  n_non_stop_words  \
0            12.0             219.0         0.663594               1.0   
1             9.0             255.0         0.604743               1.0   
2             9.0             211.0         0.575130               1.0   
3             9.0             531.0         0.503788               1.0   
4            13.0            1072.0         0.415646               1.0   

   n_non_stop_unique_tokens  num_hrefs  num_self_hrefs  num_imgs  ...  \
0                  0.815385        4.0             2.0       1.0  ...   
1                  0.791946        3.0             1.0       1.0  ...   
2                  0.663866        3.0             1.0       1.0  ...   
3                  0.665635        9.0             0.0       1.0  ...   
4                  0.540890       19.0            19.0      20.0  ...   

   avg_negative_polarity  min_negative_polarity  max_negative_polarity  \
0              -0.350000                 -0.600              -0.200000   
1              -0.118750                 -0.125              -0.100000   
2              -0.466667                 -0.800              -0.133333   
3              -0.369697                 -0.600              -0.166667   
4              -0.220192                 -0.500              -0.050000   

   title_subjectivity  title_sentiment_polarity  abs_title_subjectivity  \
0            0.500000                 -0.187500                0.000000   
1            0.000000                  0.000000                0.500000   
2            0.000000                  0.000000                0.500000   
3            0.000000                  0.000000                0.500000   
4            0.454545                  0.136364                0.045455   

   abs_title_sentiment_polarity  shares       category     Day  
0                      0.187500     593  Entertainment  Monday  
1                      0.000000     711       business  Monday  
2                      0.000000    1500       business  Monday  
3                      0.000000    1200  Entertainment  Monday  
4                      0.136364     505           tech  Monday  

[5 rows x 49 columns]

In [72]:
df=df[['category','Day','n_tokens_title','n_tokens_content','num_hrefs','num_imgs','avg_positive_polarity','avg_negative_polarity','global_subjectivity']]

In [73]:
df

category        Day  n_tokens_title  n_tokens_content  num_hrefs  \
0      Entertainment     Monday            12.0             219.0        4.0   
1           business     Monday             9.0             255.0        3.0   
2           business     Monday             9.0             211.0        3.0   
3      Entertainment     Monday             9.0             531.0        9.0   
4               tech     Monday            13.0            1072.0       19.0   
...              ...        ...             ...               ...        ...   
39639           tech  Wednesday            11.0             346.0        9.0   
39640         socmed  Wednesday            12.0             328.0        9.0   
39641      lifestyle  Wednesday            10.0             442.0       24.0   
39642          World  Wednesday             6.0             682.0       10.0   
39643  Entertainment  Wednesday            10.0             157.0        1.0   

       num_imgs  avg_positive_polarity  avg_negative_polarity  \
0           1.0               0.378636              -0.350000   
1           1.0               0.286915              -0.118750   
2           1.0               0.495833              -0.466667   
3           1.0               0.385965              -0.369697   
4          20.0               0.411127              -0.220192   
...         ...                    ...                    ...   
39639       1.0               0.333791              -0.260000   
39640       3.0               0.374825              -0.211111   
39641      12.0               0.307273              -0.356439   
39642       1.0               0.236851              -0.205246   
39643       0.0               0.247338              -0.200000   

       global_subjectivity  
0                 0.521617  
1                 0.341246  
2                 0.702222  
3                 0.429850  
4                 0.513502  
...                    ...  
39639             0.482679  
39640             0.564374  
39641             0.510296  
39642             0.358578  
39643             0.517893  

[38463 rows x 9 columns]

In [74]:
df.to_csv("ARM_Data.csv")

In [75]:
df.describe()

n_tokens_title  n_tokens_content     num_hrefs      num_imgs  \
count    38463.000000      38463.000000  38463.000000  38463.000000   
mean        10.382419        563.295375     11.217872      4.563061   
std          2.113800        468.299538     11.340580      8.295365   
min          2.000000         18.000000      0.000000      0.000000   
25%          9.000000        259.000000      5.000000      1.000000   
50%         10.000000        423.000000      8.000000      1.000000   
75%         12.000000        729.000000     14.000000      4.000000   
max         23.000000       8474.000000    304.000000    128.000000   

       avg_positive_polarity  avg_negative_polarity  global_subjectivity  
count           38463.000000           38463.000000         38463.000000  
mean                0.364689              -0.267493             0.456984  
std                 0.085455               0.121174             0.088386  
min                 0.000000              -1.000000             0.000000  
25%                 0.311880              -0.331532             0.402457  
50%                 0.361872              -0.257738             0.456566  
75%                 0.413254              -0.193415             0.510305  
max                 1.000000               0.000000             1.000000

There are some columns in dataset and these columns are numerical with diffrent ranges. columns are (n_tokens_title with minimum =2 and maximum=23),(n_tokens_content with minimum=18 and maximum=8474),(num_hrefs with minimum =0 and maximum=304.000000),(num_imgs with minimum =0 and maximum=128),(avg_positive_polarity with minimum =0 and maximum=1.000000, mean=0.364689, standard deviation=0.085455),(avg_negative_polarity with minimum =-1.000000 and maximum=0.000000, mean=-0.267493, standard deviation=0.121174), (global_subjectivity with minimum =0.000000 and maximum=1.000000, mean=0.456984, standard deviation=0.088386)

In [77]:
import pandas as pd
import numpy as np
#from mlxtend.preprocessing import TransactionEncoder

# Load the dataset into a pandas dataframe

# Define the bin sizes for each column
bins_n_tokens_title = [0, 7, 12, 17, 23]
bins_n_tokens_content = [0, 300, 600, 900, 1200, 8474]
bins_num_hrefs = [-1, 10, 20, 30, 40, 304]
bins_num_imgs = [-1, 10, 25, 50, 75, 128]
bins_avg_positive_polarity = [0, 0.2, 0.4, 0.6, 0.8, 1.0]
bins_avg_negative_polarity = [-2, -0.8, -0.6, -0.4, -0.2, 0.0]
bins_global_subjectivity = [0, 0.2, 0.4, 0.6, 0.8, 1.0]

# Discretize each column and create a new categorical column
df['n_tokens_title_cat'] = pd.cut(df['n_tokens_title'], bins=bins_n_tokens_title, labels=['Very Short Title', 'Short Title', 'Medium Title', 'Long Title'])
df['n_tokens_content_cat'] = pd.cut(df['n_tokens_content'], bins=bins_n_tokens_content, labels=['Very Short Content', 'Short Content', 'Medium Content', 'Long Content', 'Very Long Content'])
df['num_hrefs_cat'] = pd.cut(df['num_hrefs'], bins=bins_num_hrefs, labels=['Very Few Links', 'Few Links', 'Moderate Links', 'Many Links', 'Very Many Links'])
df['num_imgs_cat'] = pd.cut(df['num_imgs'], bins=bins_num_imgs, labels=['Very Few Images', 'Few Images', 'Moderate Images', 'Many Images', 'Very Many Images'])
df['avg_positive_polarity_cat'] = pd.cut(df['avg_positive_polarity'], bins=bins_avg_positive_polarity, labels=['Very Low Positive Polarity', 'Low Positive Polarity', 'Moderate Positive Polarity', 'High Positive Polarity', 'Very High Positive Polarity'])
df['avg_negative_polarity_cat'] = pd.cut(df['avg_negative_polarity'], bins=bins_avg_negative_polarity, labels=['Very High Negative Polarity', 'High Negative Polarity', 'Moderate Negative Polarity', 'Low Negative Polarity', 'Very Low Negative Polarity'])
df['global_subjectivity_cat'] = pd.cut(df['global_subjectivity'], bins=bins_global_subjectivity, labels=['Very Low Subjectivity', 'Low Subjectivity', 'Moderate Subjectivity', 'High Subjectivity', 'Very High Subjectivity'])

# Create a list of lists where each inner list is a transaction

# Use TransactionEncoder to convert the list of lists to a transaction dataframe
#te = TransactionEncoder()
#te_ary = te.fit(transactions).transform(transactions)
#df_tf = pd.DataFrame(te_ary, columns=te.columns_)

# Save the transaction dataframe to a file
#df_tf.to_csv('transactions.csv', index=False)



In [85]:
df.drop(columns=['n_tokens_title','n_tokens_content','num_hrefs','num_imgs','avg_positive_polarity','avg_negative_polarity','global_subjectivity'], inplace=True)

In [80]:
transactions = df[['category','Day','n_tokens_title_cat', 'n_tokens_content_cat', 'num_hrefs_cat', 'num_imgs_cat', 'avg_positive_polarity_cat', 'avg_negative_polarity_cat', 'global_subjectivity_cat']].values.tolist()

In [81]:
transactions

[['Entertainment',
  'Monday',
  'Short Title',
  'Very Short Content',
  'Very Few Links',
  'Very Few Images',
  'Low Positive Polarity',
  'Low Negative Polarity',
  'Moderate Subjectivity'],
 ['business',
  'Monday',
  'Short Title',
  'Very Short Content',
  'Very Few Links',
  'Very Few Images',
  'Low Positive Polarity',
  'Very Low Negative Polarity',
  'Low Subjectivity'],
 ['business',
  'Monday',
  'Short Title',
  'Very Short Content',
  'Very Few Links',
  'Very Few Images',
  'Moderate Positive Polarity',
  'Moderate Negative Polarity',
  'High Subjectivity'],
 ['Entertainment',
  'Monday',
  'Short Title',
  'Short Content',
  'Very Few Links',
  'Very Few Images',
  'Low Positive Polarity',
  'Low Negative Polarity',
  'Moderate Subjectivity'],
 ['tech',
  'Monday',
  'Medium Title',
  'Long Content',
  'Few Links',
  'Few Images',
  'Moderate Positive Polarity',
  'Low Negative Polarity',
  'Moderate Subjectivity'],
 ['tech',
  'Monday',
  'Short Title',
  'Short Conte

In [83]:
import csv

# Save transaction list as CSV file
with open('transactions.csv', 'w', newline='') as file:
    writer = csv.writer(file)
    writer.writerows(transactions)

In [86]:
df.head()

category     Day n_tokens_title_cat n_tokens_content_cat  \
0  Entertainment  Monday        Short Title   Very Short Content   
1       business  Monday        Short Title   Very Short Content   
2       business  Monday        Short Title   Very Short Content   
3  Entertainment  Monday        Short Title        Short Content   
4           tech  Monday       Medium Title         Long Content   

    num_hrefs_cat     num_imgs_cat   avg_positive_polarity_cat  \
0  Very Few Links  Very Few Images       Low Positive Polarity   
1  Very Few Links  Very Few Images       Low Positive Polarity   
2  Very Few Links  Very Few Images  Moderate Positive Polarity   
3  Very Few Links  Very Few Images       Low Positive Polarity   
4       Few Links       Few Images  Moderate Positive Polarity   

    avg_negative_polarity_cat global_subjectivity_cat  
0       Low Negative Polarity   Moderate Subjectivity  
1  Very Low Negative Polarity        Low Subjectivity  
2  Moderate Negative Polarity       High Subjectivity  
3       Low Negative Polarity   Moderate Subjectivity  
4       Low Negative Polarity   Moderate Subjectivity

In [87]:
transactions=pd.read_csv("transactions.csv")

In [89]:
transactions.head()

Entertainment  Monday   Short Title  Very Short Content  Very Few Links  \
0       business  Monday   Short Title  Very Short Content  Very Few Links   
1       business  Monday   Short Title  Very Short Content  Very Few Links   
2  Entertainment  Monday   Short Title       Short Content  Very Few Links   
3           tech  Monday  Medium Title        Long Content       Few Links   
4           tech  Monday   Short Title       Short Content  Very Few Links   

   Very Few Images       Low Positive Polarity       Low Negative Polarity  \
0  Very Few Images       Low Positive Polarity  Very Low Negative Polarity   
1  Very Few Images  Moderate Positive Polarity  Moderate Negative Polarity   
2  Very Few Images       Low Positive Polarity       Low Negative Polarity   
3       Few Images  Moderate Positive Polarity       Low Negative Polarity   
4  Very Few Images       Low Positive Polarity  Very Low Negative Polarity   

   Moderate Subjectivity  
0       Low Subjectivity  
1      High Subjectivity  
2  Moderate Subjectivity  
3  Moderate Subjectivity  
4  Moderate Subjectivity